### DISCUSSION TITLE
how to use "router.push " outside vue component in the js class?

### DISCUSSION THREAD

**Discussion by nekooee** [quasarframework/quasar#14676] [CLI - PWA mode]
Hi,

I used the latest version of the quasar framework in CLI mode. (V2.9.2)
 my index.js for route code:

```js
import { route } from "quasar/wrappers";
import {
  createRouter,
  createMemoryHistory,
  createWebHistory,
  createWebHashHistory,
} from "vue-router";
import routes from "./routes";

export default route(function (/* { store, ssrContext } */) {
  const createHistory = process.env.SERVER
    ? createMemoryHistory
    : process.env.VUE_ROUTER_MODE === "history"
    ? createWebHistory
    : createWebHashHistory;

  const Router = createRouter({
    scrollBehavior: () => ({ left: 0, top: 0 }),
    routes,
    history: createHistory(process.env.VUE_ROUTER_BASE),
  });

  return Router;
});
```

and my class in "src/js" folder:
```js
import { api } from "boot/axios";

class Auth {
  logout() {
    localStorage.clear();
    localStorage.removeItem("token");
    localStorage.removeItem("user");
    this.user = null;
    router.push({ name: "login" });
  }
}
export default new Auth();
```

but router.push doesn't work. how to use router push outside vue component?

**Comment by Junvary** [comment]
try this:
```javascript
import { useRouter } from 'vue-router';

const router = useRouter();
router.push({ name: "login" });
```

**Reply by nekooee** [reply]
> try this:
> 
> ```js
> import { useRouter } from 'vue-router';
> 
> const router = useRouter();
> router.push({ name: "login" });
> ```

No, this code only works correctly when written in a vue component.  It gives the following error outside the component:
```
 inject() can only be used inside setup() or functional components
TypeError: Cannot read properties of undefined (reading 'push')

```

**Reply by metalsadman** [reply]
export your router instance and import it where you want to use it outside of vue files.

**Reply by marty0678** [reply]
I know this is an old thread - but doing that breaks HMR, even if you aren't touching the router or route code. Anywhere that your exported router is used it starts throwing `Cannot read properties of null`. Been trying to find a work around that keeps HMR.

Update - This actually appears to work: https://dev.to/davidbuc/how-to-use-router-inside-axios-interceptors-react-and-vue-5ece
